feat: Make snapshotID optional for volumeInitializationRate when AMI is selected via amiSelectorTerms

atilsensalduz opened the pull request:
Fixes #8389 

**Description**

This PR implements the enhancement requested in #8389 to make `snapshotID` optional for `volumeInitializationRate` when AMI is selected via `amiSelectorTerms`. This enables users to benefit from faster EBS volume initialization while using dynamic AMI selection, reducing operational overhead.

Currently, Karpenter requires an explicit `snapshotID` when using `volumeInitializationRate` in block device mappings. This prevents dynamic AMI selection via `amiSelectorTerms` while still benefiting from faster EBS volume initialization, creating operational overhead for users who want to manage AMIs dynamically.

## Solution

Updated the CEL validation rule to allow `volumeInitializationRate` without explicit `snapshotID` when the EC2NodeClass has `amiSelectorTerms` configured. When AMIs are selected dynamically, they contain block device mapping information with snapshots that AWS can use for volume initialization.

**How was this change tested?**
```
go test ./pkg/apis/v1 -run "TestCEL.*BlockDeviceMappings" -v
```

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates <!-- docs must be added to /preview to be included in future version releases -->
- [ ] Yes, issue opened: # <!-- issue number -->
- [x] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

jmdeal commented:
Do we need this validation at all? Users always need to specify via `amiSelectorTerms`, there's no other way to select an AMI in Karpenter. The one edge-case where I could see this not being supported is when you're using an AMI which uses the instance's local storage rather than EBS as the root volume ([ref](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/creating-an-ami-instance-store.html)). There's no way for us to detect the type of AMI selected at validation time, so we can't perform a CEL validation against that.

Given so few instances that support that edge-case anyway and you would have to explicitly opt-in to encounter an issue (by configuring volumeInitializationRate) it seems reasonable to not perform this validation at all. I'd need to test locally to see if there are any other edge cases I'm unaware of - I'm not an expert on this EBS feature.

jmdeal commented:
Actually, there is one other thing we need to handle here - even though the snapshot ID for the root device is derived implicitly from the AMI, you still can't create a launch template that only specifies `VolumeInitializationRate` and use it with `ec2:RunInstances` - you get and `InvalidParameterCombination` error. I imagine all we should need to do is copy the snapshot ID for the AMI's root block device (discovered via DescribeImages) and pipe that into our generated launch template. That shouldn't be too difficult but does expand the scope of this PR.

moko-poi commented:
Hey @atilsensalduz,

Just checking in on this PR - any thoughts on @jmdeal's feedback from December? Particularly regarding the `InvalidParameterCombination` error and potentially piping the snapshot ID from the AMI's root block device?

Let me know if you're still working on this or need any assistance!

atilsensalduz commented:
Hi @jmdeal, apologies for missing your earlier comments.

You're right. The current Karpenter version mandates amiSelectorTerms, so the CEL validation would always pass. So, I've removed it.

I've also adjusted the PR based on your suggestion. Right now, if a user specifies volumeInitializationRate on their root volume without providing an explicit snapshotID, we now inject the AMI's snapshot ID automatically. This way users don't need to manually look up and specify the snapshot ID anymore. Karpenter handles it transparently.

Thank you so much for the heads-up about the InvalidParameterCombination error. That was really helpful!

Let me know if you have any concerns or if I've missed anything else.

atilsensalduz commented:
Hi @moko-poi , thank you for catching this. I've addressed @jmdeal's feedbacks in the latest changes. Could you please take a look at the recent changes when you get a chance?

Sarthug99 commented on pkg/apis/v1/ec2nodeclass_status.go:
```diff
@@ -77,6 +77,12 @@ type AMI struct {
 	// Requirements of the AMI to be utilized on an instance type
 	// +required
 	Requirements []corev1.NodeSelectorRequirement `json:"requirements"`
+	// RootDeviceName is the device name of the root volume (/dev/xvda)
```
nit : add eg. here matching the CRD description. /dev/xvda isn't the only possible root device name.

Sarthug99 commented:
One more thing to consider here - 
The old CEL rule rejected `volumeInitializationRate` without snapshot ID on **any** block device mapping. The new injection logic only covers root volumes (matched by device name or rootVolume flag). 
This means a non-root volume with `volumeInitializationRate` and no snapshot ID will now pass the validation and fail at the EC2 API. We should either keep the CEL validation for non-root volumes, or add validation in `injectSnapshotIDForVolumeInitialization` to address this scenario.
